components/org.wso2.carbon.identity.api.user.recovery/org.wso2.carbon.identity.rest.api.user.recovery.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/recovery/v1/impl/core/PasswordRecoveryService.java

```diff
@@ -0,0 +1,466 @@
+     *                     information
+     * @return Response
+     */
+    public Response resetPassword(ResetRequest resetRequest) {
+
+        String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+        char[] password = resetRequest.getPassword().toCharArray();
+        try {
+            SuccessfulPasswordResetDTO successfulPasswordResetDTO =
+                    UserAccountRecoveryServiceDataHolder.getPasswordRecoveryManager()
+                            .reset(resetRequest.getResetCode(), password,
+                                    RecoveryUtil.buildPropertiesMap(resetRequest.getProperties()));
+            return Response.ok().entity(buildPasswordResetResponse(successfulPasswordResetDTO)).build();
+        } catch (IdentityRecoveryClientException e) {
+            // Send the reset code again for a retry attempt.
+            throw RecoveryUtil.handleClientExceptions(PasswordRecoveryService.class.getName(), tenantDomain,
+                    IdentityRecoveryConstants.PASSWORD_RECOVERY_SCENARIO, resetRequest.getResetCode(),
+                    Util.getCorrelation(), e);
+        } catch (IdentityRecoveryException e) {
+            throw RecoveryUtil.buildInternalServerErrorResponse(PasswordRecoveryService.class.getName(),
+                    Constants.SERVER_ERROR, e.getErrorCode(), Util.getCorrelation(), e);
+        } catch (Throwable throwable) {
+            throw RecoveryUtil
+                    .buildInternalServerErrorResponse(PasswordRecoveryService.class.getName(), Constants.SERVER_ERROR,
+                            IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_UNEXPECTED_ERROR.getCode(),
+                            Util.getCorrelation(), throwable);
+        }
+    }
+
+    /**
+     * Resend the recovery confirmation code to the user.
+     *
+     * @param resendConfirmationRequest ResendConfirmationRequest {@link ResendConfirmationRequest} object
+     *                                  which wraps the resend request
+     * @return Response
+     */
+    public Response resendConfirmation(ResendConfirmationRequest resendConfirmationRequest) {
+
+        String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+        String resendCode = resendConfirmationRequest.getResendCode();
+        Map<String, String> properties = RecoveryUtil.buildPropertiesMap(resendConfirmationRequest.getProperties());
+        try {
+            ResendConfirmationDTO resendConfirmationDTO =
+                    UserAccountRecoveryServiceDataHolder.getPasswordRecoveryManager()
+                            .resend(tenantDomain, resendCode, properties);
+            if (resendConfirmationDTO == null) {
+                if (log.isDebugEnabled()) {
+                    log.debug("No ResendConfirmationDTO data for resend code :" + resendCode);
+                }
+                return Response.status(Response.Status.OK).build();
```

Comment:
If the code to reach this line without any errors, that means recovery is successful but reponse object is no there.
Using 404 is wrong since the 404 error status code indicates that the REST API can’t map the client’s URI to a resource but may be available in the future.

Therefore 500 will be a better option. WDYT?